### ISSUE TITLE
Issues with node-pre-gyp

### ISSUE THREAD

**Issue by onebesky** [cloudfoundry/nodejs-buildpack#186]
Starting version v1.6.46 we noticed that node-pre-gyp is missing in the buildpack. A dependency NPM package "canvas" that requires the package now fails with the following error:

```
error /tmp/app/node_modules/canvas: Command failed.
   Exit code: 127
   Command: node-pre-gyp install --fallback-to-build
   Arguments: 
   Directory: /tmp/app/node_modules/canvas
   Output:
   /bin/sh: 1: node-pre-gyp: not found
   info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
          **ERROR** Unable to build dependencies: exit status 127
   Cell a827573d-ee33-414e-bebd-3532a9afb61e stopping instance 84054ab3-4761-4fc2-9d78-f8314d329c4e
   Exit status 143
```
Our node_modules are not vendored and are installed on the server with cf push command. This happens for both cflinuxfs2 and cflinuxcf3.

Please confirm where necessary:
* [X] I have included a log output
* [X] My log includes an error message
* [X] I have included steps for reproduction

**Comment by tylerphelan** [comment]
@onebesky add it as a dependency to your `package.json` before `canvas`:
```
"dependencies": {
    "node-pre-gyp": "*",
    "canvas": "^2.4.1"
  }
```

**Comment by Zanadar** [comment]
Closing. Workaround provided. Please reopen if this doesn't fix your issue.

**Comment by vinsonchuong** [comment]
@tylerphelan and @Zanadar, the above solution does not solve the problem for me.

I put in some cycles to track down what was going on. Here's a reproduction example that mimics the command the buildpack uses to install dependencies

https://github.com/cloudfoundry/nodejs-buildpack/blob/b234a4553c4e680d488b54e05dbf0e967454d575/src/nodejs/yarn/yarn.go#L31

```sh
docker run -it node:latest /bin/bash

mkdir /root/project
cd /root/project
yarn add bcrypt
rm -rf node_modules
yarn install --modules-folder /root/project/node_modules
```

`bcrypt` is a package that runs `node-pre-gyp` in its `install` script (just like `canvas`). They expect `node-pre-gyp` to be on the path. `yarn` does indeed install `node-pre-gyp` correctly, but it looks like passing in `--modules-folder /root/project/node_modules` mangles the `PATH` environment variable.

In the example above, by the time `bcrypt` attempts to execute `node-pre-gyp`, it's `PATH` will include `/root/project/node_modules/bcrypt/root/project/node_modules/.bin`, which is wrong; it should instead include `/root/project/node_modules/bcrypt/node_modules/.bin`

Long story short, it looks like yarn has a bug when using `--modules-folder`

Is it possible for you to stop using that flag? As it stands I have to deploy using v1.6.45 of the buildpack.

I'll pester the yarn folks about this separately.

**Comment by vinsonchuong** [comment]
I forked the buildpack and can confirm that my app deploys without the flag:

https://github.com/vinsonchuong/nodejs-buildpack/commit/5fe64aaffc5b28d521e0332662932f1d4fbc63e9

**Comment by tylerphelan** [comment]
@vinsonchuong

**Comment by tylerphelan** [comment]
@vinsonchuong thanks for responding to this! This should be supported for yarn and we will prioritize it soon!

**Comment by mbyrne00** [comment]
Another (not-ideal) workaround until this is fixed is to remove your `yarn.lock` file since this seems to work with npm.

**Comment by shanks3012** [comment]
The fix has been pushed to the develop branch if you'd like to make use immediately. Otherwise, the fix will be included in the next release of the buildpack. Closing for now, please feel free to reopen if any issues arise.
